cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+used to generate Rust source.
+
+The decision trie is, as the name implies, a kind of decision tree, in
+the sense that we start at the root and move down the tree based on
+the result of match operations (each feeding one "decision").
+
+It is a "trie" (which is a kind of tree) because at each level, its
+edges are labeled with match operations; a trie is a tree where one
+input character from an alphabet is used to index children at each
+level.
+
+Each node in the tree is either an internal decision node, or a leaf
+"expression" node (which we reach once we have a successful rule
+match). The "execution semantics" of the trie are
+backtracking-based. We attempt to find some path down the tree through
+edges whose match ops run successfully; when we do this to reach a
+leaf, we have the values generated by all of the match ops, and we can
+execute the sequence of "expression instructions" in the leaf. Each
+rule's left-hand side becomes a series of edges (merged into the
+existing tree as we process rules) and each rule's right-hand side
+becomes one leaf node with expression instructions.
+
+At any point, if a match op does not succeed, we try the next out-edge
+in sequence. If we have tried all out-edges from a decision node and
+none were successful, then we backtrack one level further. Thus, we
+simply perform an in-order tree traversal and find the first
+successful match.
+
+Though this sounds possibly very inefficient if some decision node has
+a high fan-out, in practice it is not because the edges are often
+known to be *mutually exclusive*. The canonical example of this is
+when an enum-typed value is destructured into different variants by
+various edges; we can use a Rust `match` statement in the generated
+source and have `O(1)` (or close to it) cost for the dispatch at this
+level.
+
+Building the trie is a somewhat subtle procedure because we must be
+mindful of *priorities*. In order to ensure that rule priorities are
+obeyed, we actually label each edge with a *priority range* and a
+match op. Edges must then be considered in priority order. When
+building the trie, we uphold an important *non-overlapping-priority*
+invariant: there must not be two edges with overlapping priority
+ranges and different match ops, or else it would be unclear which to
+run next. Priority ranges are inclusive in a discrete (integer) space,
+so "overlap" is also subtle: we can allow edges with "unit" priority
+ranges (just one priority), and these do not overlap. So, e.g., we can
+allow an edge with priority `[0, 10]` (higher is more prioritized) and
+op `A`, followed by an edge with priority `[0, 0]` and op `B`,
+followed by an edge with priority `[-1, -10]`, `A`, though they must
+occur in that order.
+
+See [this block
+comment](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/isle/isle/src/trie.rs#L15-L166)
+for more information regarding the trie construction algorithm.
```

Comment:
Does this section have any information that isn't present in that comment?

This level of implementation detail seems better reserved for code comments (as linked to here) or implementation overviews (as in the README), not language reference.

I think this section can probably be removed.

Reply:
Good point, the description of the algorithm itself is probably better left to the single canonical place in the source; removed that. I left the high-level description of the trie and added a note ("We describe this data structure below with the intent to provide an understanding of how the DSL compiler...") as I think it's useful to have this mental model whenever questions of how the generated code will work come up.